### PR TITLE
chore(net): extract `NetworkHandle` methods for launching node to traits

### PR DESCRIPTION
**Pull request by emhane**
Ref https://github.com/paradigmxyz/reth/issues/9872

- Adds new traits `BlockDownloaderProvider` in order to use generic network in node builder, instead of concrete type `NetworkHandle`
- Add new trait `FullNetwork` which unifies behaviour needed to launch node
-  Adds `PeersHandleProvider` to integrate generic network in testing
- Moves necessary types from `reth-network` to `reth-network-types` in order to extract `NetworkHandle` methods into new traits

Note: In order to move `BlockDownloaderProvider` to `reth-network-api`, the `FetchClient` type would have to move to `reth-network-types`, and `reth_network_p2p::EthResponseValidator` and `reth_network_p2p::HeadersRequest` would have to move out of `reth-network-p2p`. The latter, is to remove dep on `reth-network-types` from `reth-network-p2p`, since `FetchClient` depends on `reth-network-p2p`

### CODE REVIEW

File: crates/net/network/tests/it/multiplex.rs
Comment (mattsse) [inline crates/net/network/tests/it/multiplex.rs:1]: did anything change here?
all of this looks unrelated, hard to tell

**Review by mattsse — CHANGES_REQUESTED** [review]
looks like a lot of the changes are unrelated and due to moved use statements, please try to avoid that in a pr that also moves stuff around because this makes it a lot harder to review

I'd like to break this down, can we start this process with a pr that introduces the new traits and perhaps (multiple) prs that move the types to network-types as prep?

File: crates/net/network/tests/it/multiplex.rs
Comment (emhane) [reply crates/net/network/tests/it/multiplex.rs:1]: hm, sometimes when bases switches, changes on main look like new changes. I checked this out from `emhane/network-types` originally.

Comment (emhane) [reply crates/net/network/tests/it/multiplex.rs:1]: the module preludes I crossed, I cleaned up, is this a problem?

File: crates/net/network-types/src/peers/cmd.rs
Comment (mattsse) [inline crates/net/network-types/src/peers/cmd.rs:1]: hmm, this would be more appropriate in network-api I believe

File: crates/net/network-types/src/peers/handle.rs
Comment (mattsse) [inline crates/net/network-types/src/peers/handle.rs:1]: same

File: crates/net/network-types/src/peers/mod.rs [hunk 1]
```diff
@@ -1,5 +1,141 @@
...
+use tracing::trace;
+
+use crate::{
+    is_banned_reputation, PeerAddr, PeerConnectionState, PeerKind, ReputationChangeOutcome,
+    DEFAULT_REPUTATION,
+};
+
+/// Tracks info about a single peer.
+#[derive(Debug, Clone)]
+pub struct Peer {
```
Comment (mattsse) [inline crates/net/network-types/src/peers/mod.rs:24]: why do we need to move this?

this is currently a private type

**Review by mattsse — CHANGES_REQUESTED** [review]
I would still prefer if we could do this incrementally, because it's a bit to keep an overview about why certain things are moved where.

I still think we can move the types that we need first and then introduce the traits

File: crates/net/network-types/src/peers/mod.rs (hunk 1 above)
Comment (emhane) [reply crates/net/network-types/src/peers/mod.rs:24]: can move this back to network

File: crates/net/network-types/src/peers/handle.rs
Comment (emhane) [reply crates/net/network-types/src/peers/handle.rs:1]: this is moved so that `PeersHandleProvider` can be moved to `reth-network-api`

File: crates/net/network-types/src/peers/cmd.rs
Comment (emhane) [reply crates/net/network-types/src/peers/cmd.rs:1]: this is needed in `PeersHandle` type, so needs to stay too. we should try and keep types in the designated types crate and traits in the designated api crate, just like it's cleanly done for `reth-rpc-eth-api` and `reth-rpc-eth-types`

File: crates/net/network-types/src/peers/mod.rs (hunk 1 above)
Comment (emhane) [reply crates/net/network-types/src/peers/mod.rs:24]: it's a public type on main though

Comment (emhane) [reply crates/net/network-types/src/peers/mod.rs:24]: ah no I can't move it back, because it's used in `PeersHandle` impl

**Comment by emhane** [comment]
> I would still prefer if we could do this incrementally, because it's a bit to keep an overview about why certain things are moved where.
> 
> I still think we can move the types that we need first and then introduce the traits

I think that argument is contradictory, since the traits and cyclical deps are the reason why the types need to move

File: crates/net/network-types/src/peers/cmd.rs
Comment (mattsse) [reply crates/net/network-types/src/peers/cmd.rs:1]: these are all network api related types so I don't understand why we can't move this to the API crate, types is intended for standalone network related types, types we only used for interacting with network API functions don't fit here like command types

Comment (emhane) [reply crates/net/network-types/src/peers/cmd.rs:1]: we can re-export them in `reth-network-api`, but it's much more intentional design if we keep types in types and api in api

**Review by mattsse — APPROVED** [review]
okay convinced but please keep these smaller moving forward
